### DISCUSSION TITLE
IPC Luajit implementation

### DISCUSSION THREAD

**Discussion by A-Cloud-Ninja** [JakeStanger/ironbar#235] [Show and tell]
Dependencies:
[1-file JSON library from rxi](https://github.com/rxi/json.lua)
[Luajit](https://luajit.org/) or another compatible interpreter with a luajit-compatible FFI syntax (maybe [luaffi](https://github.com/jmckaskill/luaffi))
[ironbar.lua](https://gist.github.com/A-Cloud-Ninja/4f2921a07fe0f89d0f6c8428fcb6c5be)
I've written a very basic implementation of ironbars IPC spec in luajit, it provides all current (July 10th 2023) IPC calls and the responses.

It isn't particularly great, doesn't do a lot of error checking, but functionally it's all there and should be easy to expand moving forward.
Basic usage:
- `require("ironbar")` Returns a function to obtain both the ironbar library, and the json library passed through.
```lua
local ironbar,json = require("ironbar")()
ironbar.ping()
ironbar.get("key")
ironbar.set("key","value")
ironbar.load_css("/path/to/stylesheet.css")
--I also built some goofy syntax for using it this way
ironbar("ping")
ironbar("get","key")
ironbar("set","key","value")
ironbar("load_css","/path/to/stylesheet.css")
```
All of these will return the decoded json as a table for use.

**Comment by yavko** [comment]
This is epic! This makes me want to copy the built-in code and make it into a rust library, lol

**Reply by A-Cloud-Ninja** [reply]
If you do, I implore you to look at potentially making it truly native. Then in theory, all any other language would have to do is write some minor glue around hopefully a plain `dispatch` function like I have setup for the goofy metatable syntax, rather than needing to rewrite a unix socket header in cdefs haha.

**Reply by yavko** [reply]
Rust is truly native, problem is that it doesn't have stable ABI, i could make a wrapper over the library that compiled to a dylib, and used C types though

**Reply by A-Cloud-Ninja** [reply]
That sounds like the perfect option!

**Reply by JakeStanger** [reply]
Having something based in C would be great, as that makes it trivial to port to like *everything* else. It sounds like you know more about that than I do, so if you're up for it absolutely go for it :)

**Reply by yavko** [reply]
I have an idea, instead of keeping the ipc client in ironbar itself, could you create a sub crate that holds the ipc client, then use it from ironbar. This would enable a rust ipc client library without extra work.

**Reply by JakeStanger** [reply]
Yeah I was considering that - I can split the repo up into a workspace and have the IPC & types in their own crate, then we can stick a third client library crate alongside those. 

It would involve moving all the source, so I need anything outstanding to be merged first to avoid completely destruction (discluding the GTK4 PR because that's a billion miles off still)

**Reply by yavko** [reply]
Haha, I guess that includes me. Sadly, I haven't had time to work on the PR, but I will in a couple of days.

**Comment by A-Cloud-Ninja** [comment]
Update: I've moved this to its' own repo as it's gotten large enough to warrant separating out some of the boilerplate
https://github.com/A-Cloud-Ninja/ironbar-ipc-luajit

**Reply by JakeStanger** [reply]
Awesome stuff, thanks! I'll add a link in the wiki to this shortly

**Comment by A-Cloud-Ninja** [comment]
> Rust is truly native, problem is that it doesn't have stable ABI, i could make a wrapper over the library that compiled to a dylib, and used C types though

@yavko 
Currently working on a small piece for the libcorn repo, [rust *is* in fact stable enough to export a lua lib with mlua!](https://kdheepak.com/blog/loading-a-rust-library-as-a-lua-module-in-neovim/)

**Reply by yavko** [reply]
oh thats sick

**Reply by A-Cloud-Ninja** [reply]
see https://github.com/JakeStanger/corn/pull/20 for use of this in practice too
